Dc/fancy default

DanCardin opened the pull request.

pawamoy commented:
Trying it out locally. The classes (default, defaulttype, env, prompt, confirm, valuefrom) should probably implement `__str__` to allow displaying them in CLI help message. Might be tricky :thinking: For example I would like the following default:

```python
cappa.Env("SOME_ENV_VAR") | cappa.ValueFrom(from_config, "some_config_value")
```

...to display as:

```
Default: `SOME_ENV_VAR` or configuration value `some.config_value`.
```

...and there's no way to get this result automatically, of course. But I don't see how the user could provide their own string to `ValueFrom`, unless the class accepts explicit `args` and `kwargs` params to avoid name shadowing :thinking:

```python
# If `ValueFrom` accepts `helptext`, the same name cannot be used in `some_callable`.
cappa.ValueFrom(some_callable, "some_config_value", helptext="configuration value `some.config_value`")

# Instead, maybe accept this?
cappa.ValueFrom(
    some_callable,
    args=["some_config_value"],

    # These two don't clash anymore.
    kwargs={"helptext": "whatever"},
    helptext="configration value `some.config_value`",
)
```

Or maybe you have another idea on how to customize the display of defaults :relaxed:?

pawamoy commented:
What's the reasoning for ignoring exceptions raised when evaluating a default?

```python
    def __call__(self) -> tuple[bool, Any | None]:
        for default in self.sequence:
            try:
                value = default()
            except Exception:  # noqa: S112
                continue
```

I got bit by this. My `Config` object didn't declare a specific attribute, so my `from_config` callable doing `getattr(config, attr_name)` raised an exception, but it was silently ignored and the final value of the command option was `None` instead of the actual default, an empty dict built with the default factory.

pawamoy commented:
> but stuff like Env and Prompt always return strings akin to raw values from cli parsing, and should be parsed

Yes that makes sense. We must make sure it's documented in the "defaults are not parsed" paragraph :slightly_smiling_face:

DanCardin commented:
* Updated docs
* default_factory should now be lazy (by using ValueFrom)
* `parse` shouldn't be called on ValueFroms
* `State` is an object you should be able to use for sharing state rather than using a global/classvar. In your case, you should just be able to depend on it without directly constructing one.

This ended up being a much larger diff than I'd originally thought it would be, but it also simplified a bunch of the default handling for Env/Prompt and made doing default_factory properly trivial (implemented it in literally 2 minutes).

So I think it was clearly an overdue concept/refactor even if I trashed the whole State concept.

I **think** State ends up being reasonably analagous to `click.pass_context` (or at least the only reason i've ever used `pass_context` was to stash data), so I dont feel **bad** about it, but i've also never really had a reason for something like this (I always just use `Dep`s), so this bit i'm less certain about the ergonomics of.

But it **seems** like it should have utility any time someone wants to pass data from the calling context into something inside cappa's handling anyways, so 🤷

pawamoy commented:
Thank you so much! I'll try to test and report back, but do merge if you think it's ready (don't wait for me) :smile:

DanCardin commented:
I'll merge now just because it'll otherwise block other work, but i'd love if you were able to test before I cut a release

pawamoy commented:
I just read the docs, they are great! :rocket: It's a big PR indeed :joy: `State` looks cleaner than my private class variable, I'll try it too.
